File: samples/CameraAccessAndroid/app/src/main/java/com/meta/wearable/dat/externalsampleapps/cameraaccess/grok/DisplayHudManager.kt
package com.meta.wearable.dat.externalsampleapps.cameraaccess.grok

import android.util.Log
import com.meta.wearable.dat.core.Wearables
import com.meta.wearable.dat.core.selectors.SpecificDeviceSelector
import com.meta.wearable.dat.core.session.DeviceSession
import com.meta.wearable.dat.core.session.DeviceSessionState
import com.meta.wearable.dat.core.types.DeviceIdentifier
import com.meta.wearable.dat.core.types.DeviceSessionError
import com.meta.wearable.dat.display.Display
import com.meta.wearable.dat.display.addDisplay
import com.meta.wearable.dat.display.removeDisplay
import com.meta.wearable.dat.display.types.DisplayState
import com.meta.wearable.dat.display.views.ContentScope
import com.meta.wearable.dat.display.views.Direction
import com.meta.wearable.dat.display.views.FlexBoxBackground
import com.meta.wearable.dat.display.views.TextColor
import com.meta.wearable.dat.display.views.TextStyle
import com.meta.wearable.dat.externalsampleapps.cameraaccess.openclaw.ToolCallStatus
import com.meta.wearable.dat.externalsampleapps.cameraaccess.settings.SettingsManager
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Job
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch

enum class DisplayHudConnectionState(val displayText: String) {
    DISABLED("HUD Off"),
    DISCONNECTED("HUD"),
    CONNECTING("HUD..."),
    READY("HUD"),
    ERROR("HUD Error"),
}

class DisplayHudManager(private val scope: CoroutineScope) {
    companion object {
        private const val TAG = "DisplayHudManager"
    }

    private val lock = Any()
    private val _connectionState = MutableStateFlow(DisplayHudConnectionState.DISCONNECTED)
    val connectionState: StateFlow<DisplayHudConnectionState> = _connectionState.asStateFlow()

    private var session: DeviceSession? = null
    private var sharedSession: DeviceSession? = null
    private var ownsSession: Boolean = false
    private var display: Display? = null
    private var sessionStateJob: Job? = null
    private var sessionErrorJob: Job? = null
    private var displayStateJob: Job? = null
    private var pendingContent: (ContentScope.() -> Unit)? = null

    fun showStatus(title: String, body: String) {
        sendContent { card("VisionClaw", title, body) }
    }

    fun showTranscript(user: String, assistant: String) {
        if (user.isBlank() && assistant.isBlank()) return
        sendContent {
            flexBox(direction = Direction.COLUMN, gap = 10, padding = 24, background = FlexBoxBackground.CARD) {
                if (user.isNotBlank()) {
                    text("You", style = TextStyle.META, color = TextColor.SECONDARY)
                    text(clamp(user, 96), style = TextStyle.BODY)
                }
                if (assistant.isNotBlank()) {
                    text("Grok", style = TextStyle.META, color = TextColor.SECONDARY)
                    text(clamp(assistant, 140), style = TextStyle.BODY)
                }
            }
        }
    }

    fun showToolStatus(status: ToolCallStatus) {
        if (status is ToolCallStatus.Idle) return
        sendContent { card("OpenClaw", toolTitle(status), clamp(status.displayText, 120)) }
    }

    fun showVisionSummary(summary: String) {
        sendContent { card("Camera", "Visual context", clamp(summary, 150)) }
    }

    fun showModelCard(title: String, body: String, kind: String?) {
        sendContent { card((kind ?: "HUD").replaceFirstChar { it.uppercase() }, clamp(title, 40), clamp(body, 160)) }
    }

    fun showError(message: String) {
        sendContent { card("VisionClaw", "Needs attention", clamp(message, 150)) }
    }

    fun setSharedSession(deviceSession: DeviceSession?) {
        synchronized(lock) { sharedSession = deviceSession }
    }

    fun stop() {
        synchronized(lock) { pendingContent = null }
        synchronized(lock) { session }?.removeDisplay()
        cleanupDisplay()
        sessionStateJob?.cancel()
        sessionStateJob = null
        sessionErrorJob?.cancel()
        sessionErrorJob = null
        synchronized(lock) {
            if (ownsSession) {
                session?.stop()
            }
            session = null
            ownsSession = false
        }
        _connectionState.value =
            if (SettingsManager.displayHUDEnabled) DisplayHudConnectionState.DISCONNECTED
            else DisplayHudConnectionState.DISABLED
    }

    private fun sendContent(content: ContentScope.() -> Unit) {
        if (!SettingsManager.displayHUDEnabled) {
            _connectionState.value = DisplayHudConnectionState.DISABLED
            return
        }

        val currentDisplay = synchronized(lock) { display }
        if (currentDisplay != null && _connectionState.value == DisplayHudConnectionState.READY) {
            doSend(currentDisplay, content)
            return
        }

        synchronized(lock) { pendingContent = content }
        prepareDisplay()
    }

    private fun prepareDisplay() {
        val existingSession = synchronized(lock) { session }
        if (existingSession != null) {
            if (synchronized(lock) { display } == null) {
                if (existingSession.state.value == DeviceSessionState.STARTED) {
                    attachDisplay(existingSession)
                } else {
                    _connectionState.value = DisplayHudConnectionState.CONNECTING
                }
            }
            return
        }

        val currentSharedSession = synchronized(lock) { sharedSession }
        if (currentSharedSession != null) {
            synchronized(lock) {
                session = currentSharedSession
                ownsSession = false
            }
            observeSession(currentSharedSession, shouldStart = false)
            if (currentSharedSession.state.value == DeviceSessionState.STARTED) {
                attachDisplay(currentSharedSession)
            } else {
                _connectionState.value = DisplayHudConnectionState.CONNECTING
            }
            return
        }

        val deviceId = selectDisplayDevice()
        if (deviceId == null) {
            _connectionState.value = DisplayHudConnectionState.ERROR
            Log.w(TAG, "No display-capable wearable is available")
            return
        }

        _connectionState.value = DisplayHudConnectionState.CONNECTING
        Wearables.createSession(SpecificDeviceSelector(deviceId)).fold(
            onSuccess = { newSession ->
                synchronized(lock) {
                    session = newSession
                    ownsSession = true
                }
                observeSession(newSession, shouldStart = true)
            },
            onFailure = { error, _ ->
                Log.e(TAG, "Failed to create display session: ${error.description}")
                _connectionState.value = DisplayHudConnectionState.ERROR
            },
        )
    }

    private fun observeSession(currentSession: DeviceSession, shouldStart: Boolean) {
        sessionStateJob?.cancel()
        sessionErrorJob?.cancel()
        sessionStateJob = scope.launch {
            currentSession.state.collect { state ->
                when (state) {
                    DeviceSessionState.STARTED -> attachDisplay(currentSession)
                    DeviceSessionState.STOPPED -> {
                        cleanupDisplay()
                        synchronized(lock) {
                            if (session == currentSession) {
                                session = null
                                ownsSession = false
                            }
                        }
                        _connectionState.value = DisplayHudConnectionState.DISCONNECTED
                    }
                    else -> Unit
                }
            }
        }
        sessionErrorJob = scope.launch {
            currentSession.errors.collect { error -> handleSessionError(error) }
        }
        if (shouldStart) {
            currentSession.start()
        }
    }

    private fun attachDisplay(currentSession: DeviceSession) {
        if (synchronized(lock) { display } != null) return

        currentSession.addDisplay().fold(
            onSuccess = { newDisplay ->
                synchronized(lock) { display = newDisplay }
                displayStateJob = scope.launch {
                    newDisplay.state.collect { state ->
                        when (state) {
                            DisplayState.STARTED -> {
                                _connectionState.value = DisplayHudConnectionState.READY
                                val pending = synchronized(lock) {
                                    val content = pendingContent
                                    pendingContent = null
                                    content
                                }
                                if (pending != null) {
                                    doSend(newDisplay, pending)
                                }
                            }
                            DisplayState.STARTING -> _connectionState.value = DisplayHudConnectionState.CONNECTING
                            DisplayState.STOPPED -> {
                                cleanupDisplay()
                                _connectionState.value = DisplayHudConnectionState.DISCONNECTED
                            }
                            DisplayState.STOPPING -> _connectionState.value = DisplayHudConnectionState.CONNECTING
                        }
                    }
                }
            },
            onFailure = { error, _ ->
                Log.e(TAG, "Failed to attach display: ${error.description}")
                _connectionState.value = DisplayHudConnectionState.ERROR
            },
        )
    }

    private fun doSend(currentDisplay: Display, content: ContentScope.() -> Unit) {
        scope.launch {
            currentDisplay.sendContent(content).fold(
                onSuccess = {},
                onFailure = { error, _ ->
                    Log.e(TAG, "Display send failed: ${error.description}")
                    _connectionState.value = DisplayHudConnectionState.ERROR
                },
            )
        }
    }

    private fun cleanupDisplay() {
        displayStateJob?.cancel()
        displayStateJob = null
        synchronized(lock) { display = null }
    }

    private fun handleSessionError(error: DeviceSessionError) {
        Log.e(TAG, "Display session error: ${error.description}")
        _connectionState.value = DisplayHudConnectionState.ERROR
    }

    private fun selectDisplayDevice(): DeviceIdentifier? {
        val devices = Wearables.devices.value
        return devices.firstOrNull { deviceId ->
            Wearables.devicesMetadata[deviceId]?.value?.isDisplayCapable() == true
        } ?: devices.firstOrNull()
    }

    private fun ContentScope.card(label: String, title: String, body: String) {
        flexBox(direction = Direction.COLUMN, gap = 10, padding = 24, background = FlexBoxBackground.CARD) {
            text(label, style = TextStyle.META, color = TextColor.SECONDARY)
            text(title, style = TextStyle.HEADING)
            text(body, style = TextStyle.BODY)
        }
    }

    private fun toolTitle(status: ToolCallStatus): String =
        when (status) {
            is ToolCallStatus.Executing -> "Working"
            is ToolCallStatus.Completed -> "Done"
            is ToolCallStatus.Failed -> "Failed"
            is ToolCallStatus.Cancelled -> "Cancelled"
            is ToolCallStatus.Idle -> "OpenClaw"
        }

    private fun clamp(text: String, limit: Int): String {
        val trimmed = text.trim()
        return if (trimmed.length <= limit) trimmed else trimmed.take(limit) + "..."
    }
}

File: samples/CameraAccess/CameraAccess/Grok/DisplayHUDManager.swift
import Combine
import Foundation
import MWDATCore
import MWDATDisplay

enum DisplayHUDConnectionState: Equatable {
  case disabled
  case disconnected
  case connecting
  case ready
  case error(String)

  var displayText: String {
    switch self {
    case .disabled: return "HUD Off"
    case .disconnected: return "HUD"
    case .connecting: return "HUD..."
    case .ready: return "HUD"
    case .error: return "HUD Error"
    }
  }
}

@MainActor
final class DisplayHUDManager: ObservableObject {
  @Published private(set) var connectionState: DisplayHUDConnectionState = .disconnected
  @Published private(set) var lastError: String?

  private let wearables: WearablesInterface
  private var deviceSelector: AutoDeviceSelector
  private var sharedDeviceSession: DeviceSession?
  private var deviceSession: DeviceSession?
  private var ownsDeviceSession = false
  private var display: Display?
  private var stateListenerToken: AnyListenerToken?
  private var coreStateTask: Task<Void, Never>?
  private var sessionErrorTask: Task<Void, Never>?
  private var displayStateTask: Task<Void, Never>?
  private var displayStateContinuation: AsyncStream<DisplayState>.Continuation?
  private var pendingAction: (() async -> Void)?

  init(wearables: WearablesInterface) {
    self.wearables = wearables
    self.deviceSelector = AutoDeviceSelector(wearables: wearables, filter: { $0.supportsDisplay() })
  }

  deinit {
    stateListenerToken = nil
    coreStateTask?.cancel()
    sessionErrorTask?.cancel()
    displayStateTask?.cancel()
  }

  func showStatus(title: String, body: String) async {
    await send(DisplayHUDViews.status(title: title, body: body))
  }

  func showTranscript(user: String, assistant: String) async {
    guard !user.isEmpty || !assistant.isEmpty else { return }
    await send(DisplayHUDViews.transcript(user: user, assistant: assistant))
  }

  func showToolStatus(_ status: ToolCallStatus) async {
    guard status != .idle else { return }
    await send(DisplayHUDViews.toolStatus(status))
  }

  func showVisionSummary(_ summary: String) async {
    await send(DisplayHUDViews.vision(summary))
  }

  func showModelCard(title: String, body: String, kind: String?) async {
    await send(DisplayHUDViews.modelCard(title: title, body: body, kind: kind))
  }

  func showError(_ message: String) async {
    await send(DisplayHUDViews.error(message))
  }

  func useSharedDeviceSession(_ session: DeviceSession?) {
    sharedDeviceSession = session
  }

  func detachFromDisplay() async {
    stateListenerToken = nil
    displayStateContinuation?.finish()
    displayStateContinuation = nil
    displayStateTask?.cancel()
    displayStateTask = nil
    await display?.stop()
    display = nil
    coreStateTask?.cancel()
    coreStateTask = nil
    sessionErrorTask?.cancel()
    sessionErrorTask = nil
    if ownsDeviceSession {
      deviceSession?.stop()
    }
    deviceSession = nil
    ownsDeviceSession = false
    connectionState = SettingsManager.shared.displayHUDEnabled ? .disconnected : .disabled
  }

  private func send(_ view: some DisplayableView) async {
    guard SettingsManager.shared.displayHUDEnabled else {
      connectionState = .disabled
      return
    }

    if let display, connectionState == .ready {
      await doSend(view, on: display)
      return
    }

    let sendableView = view
    pendingAction = { [weak self] in
      guard let self, let display = self.display else { return }
      await self.doSend(sendableView, on: display)
    }

    if display == nil {
      await attachToDisplay()
    }
  }

  private func doSend(_ view: some DisplayableView, on capability: Display) async {
    do {
      try await capability.send(view)
    } catch {
      let message = (error as? DisplayError)?.description ?? error.localizedDescription
      lastError = message
      connectionState = .error(message)
    }
  }

  private func attachToDisplay() async {
    guard display == nil else { return }
    connectionState = .connecting
    lastError = nil

    if let sharedDeviceSession {
      deviceSession = sharedDeviceSession
      ownsDeviceSession = false
      observeDeviceSession(sharedDeviceSession)
      if sharedDeviceSession.state == .started {
        await setupDisplay(on: sharedDeviceSession)
      }
      return
    }

    do {
      let session = try wearables.createSession(deviceSelector: deviceSelector)
      deviceSession = session
      ownsDeviceSession = true
      observeDeviceSession(session)
      try session.start()
    } catch DeviceSessionError.datAppOnTheGlassesUpdateRequired {
      let message = DeviceSessionError.datAppOnTheGlassesUpdateRequired.localizedDescription
      lastError = message
      connectionState = .error(message)
    } catch {
      let message = "Failed to create display session: \(error.localizedDescription)"
      lastError = message
      connectionState = .error(message)
    }
  }

  private func observeDeviceSession(_ session: DeviceSession) {
    coreStateTask?.cancel()
    sessionErrorTask?.cancel()

    let stateStream = session.stateStream()
    let errorStream = session.errorStream()
    coreStateTask = Task { [weak self] in
      for await state in stateStream {
        guard let self, !Task.isCancelled else { return }
        switch state {
        case .started:
          await self.setupDisplay(on: session)
        case .stopping, .stopped:
          self.connectionState = .disconnected
          self.display = nil
          if self.ownsDeviceSession {
            self.deviceSession = nil
            self.ownsDeviceSession = false
          }
        case .idle, .starting, .paused:
          break
        @unknown default:
          break
        }
      }
    }
    sessionErrorTask = Task { [weak self] in
      for await error in errorStream {
        guard let self, !Task.isCancelled else { return }
        self.handleSessionError(error)
      }
    }
  }

  private func setupDisplay(on session: DeviceSession) async {
    guard display == nil else { return }

    do {
      let capability = try session.addDisplay()
      let (stateStream, continuation) = AsyncStream.makeStream(of: DisplayState.self)
      displayStateContinuation = continuation
      stateListenerToken = capability.statePublisher.listen { state in
        continuation.yield(state)
      }

      displayStateTask = Task { [weak self] in
        for await state in stateStream {
          guard let self, !Task.isCancelled else { return }
          switch state {
          case .starting:
            self.connectionState = .connecting
          case .started:
            self.connectionState = .ready
            if let action = self.pendingAction {
              self.pendingAction = nil
              await action()
            }
          case .stopping:
            self.connectionState = .connecting
          case .stopped:
            self.connectionState = .disconnected
            self.stateListenerToken = nil
            self.displayStateContinuation?.finish()
            self.displayStateContinuation = nil
            self.display = nil
          }
        }
      }

      await capability.start()
      display = capability
    } catch {
      let message = "Failed to start display: \(error.localizedDescription)"
      lastError = message
      connectionState = .error(message)
    }
  }

  private func handleSessionError(_ error: DeviceSessionError) {
    let message = error.localizedDescription
    lastError = message
    connectionState = .error(message)
  }
}

private enum DisplayHUDViews {
  static func status(title: String, body: String) -> FlexBox {
    base(label: "VisionClaw", title: title, body: body)
  }

  static func transcript(user: String, assistant: String) -> FlexBox {
    FlexBox(direction: .column, spacing: 10) {
      if !user.isEmpty {
        Text("You", style: .meta, color: .secondary)
        Text(clamp(user, limit: 96), style: .body)
      }
      if !assistant.isEmpty {
        Text("Grok", style: .meta, color: .secondary)
        Text(clamp(assistant, limit: 140), style: .body)
      }
    }
    .padding(24)
    .background(.card)
  }

  static func toolStatus(_ status: ToolCallStatus) -> FlexBox {
    base(label: "OpenClaw", title: toolTitle(status), body: clamp(status.displayText, limit: 120))
  }

  static func vision(_ summary: String) -> FlexBox {
    base(label: "Camera", title: "Visual context", body: clamp(summary, limit: 150))
  }

  static func modelCard(title: String, body: String, kind: String?) -> FlexBox {
    base(label: (kind ?? "HUD").capitalized, title: clamp(title, limit: 40), body: clamp(body, limit: 160))
  }

  static func error(_ message: String) -> FlexBox {
    base(label: "VisionClaw", title: "Needs attention", body: clamp(message, limit: 150))
  }

  private static func base(label: String, title: String, body: String) -> FlexBox {
    FlexBox(direction: .column, spacing: 10) {
      Text(label, style: .meta, color: .secondary)
      Text(title, style: .heading)
      Text(body, style: .body)
    }
    .padding(24)
    .background(.card)
  }

  private static func toolTitle(_ status: ToolCallStatus) -> String {
    switch status {
    case .executing: return "Working"
    case .completed: return "Done"
    case .failed: return "Failed"
    case .cancelled: return "Cancelled"
    case .idle: return "OpenClaw"
    }
  }

  private static func clamp(_ text: String, limit: Int) -> String {
    let trimmed = text.trimmingCharacters(in: .whitespacesAndNewlines)
    guard trimmed.count > limit else { return trimmed }
    let index = trimmed.index(trimmed.startIndex, offsetBy: limit)
    return String(trimmed[..<index]) + "..."
  }
}
